### ISSUE TITLE
Change "Scenario" in "Benchmark" Object

### ISSUE THREAD

**Issue by vlomonaco** [ContinualAI/avalanche#503]
Following the nomenclature described in the paper I think we should remove the notion of "Scenario" from the codebase. 

At the moment, having benchmarks generators that return scenario objects is not intuitive. It is fine to use the word scenario to describe a particular setting (NC scenario, Task-Incremental scenario, etc..). However, in the classes and methods names I think this is not necessary. For example, it would be better to have `NCBenchmark` objects rather than `NCScenario` objects that offer the stream of exps.

The same goes for the generators: "nc_scenario" -> "nc_benchmark".

What do you think?

**Comment by lrzpellegrini** [comment]
Makes sense. However, there is a practical implication regarding the name of the `avalanche.benchmarks.scenarios` package, which I don't know how to rename. Any suggestion?

**Comment by lrzpellegrini** [comment]
Hi, I'm in the process of updating helper methods to implement the support for custom streams. While doing so I'mm probably tackle this issue too. See #519

**Comment by vlomonaco** [comment]
ok perfect!

**Comment by vlomonaco** [comment]
This is half-way through, we should complete it asap!

**Comment by mmasana** [comment]
To add to this discussion from the meeting. I was also a bit confused about the nomenclature for both. My understanding was that a scenario explains the set of assumptions (for example, class-incremental learning includes learning a sequence of tasks without task-ID), while a benchmark is a specific combination of a scenario with a dataset and other possible "rules" (hyperparameters or restrictions) to make a fair comparison between different strategies/plugins (also confusing nomenclature to me). Therefore, experiments would be run on a specific benchmark (CIFAR-100 with an exemplar memory of 2,000 exemplars) for a certain scenario (class-incremental learning).

How will this look like after the change to only use benchmark (removing scenario)?

**Comment by vlomonaco** [comment]
Hi @mmasana!

> I was also a bit confused about the nomenclature for both. My understanding was that a scenario explains the set of assumptions (for example, class-incremental learning includes learning a sequence of tasks without task-ID), while a benchmark is a specific combination of a scenario with a dataset and other possible "rules" (hyperparameters or restrictions).

This is correct. The confusing aspect was mostly about the fact that when you create a benchmark the actual python object was not an instance of a "benchmark" class but was actually an object instance of the "scenario" class. This was confusing for some people so we decided to have a benchmark class that given a dataset generates a benchmark instance. Scenario is a word that can be used to describe theoretically a specific settings (e.g. "class-incremental") but it's not necessary in the code since we do not impose a strict classification but we offer only benchmarks generators that can be useful to suit some scenarios, yes, but in the end return a benchmark instance. Does it make sense? :)

> [...] to make a fair comparison between different strategies/plugins (also confusing nomenclature to me).
Why do you say strategy/plugins is a confusing nomenclature? Can you elaborate on this? :)

**Comment by mmasana** [comment]
Thanks for the answer @vlomonaco! Then with the modifications proposed yesterday on the documentation, it will be clearer. Are you also planning on giving some examples in the tutorials? Like showing a list of classic benchmarks for each scenario? I think that could be useful.

> Why do you say strategy/plugins is a confusing nomenclature? Can you elaborate on this? :)

The nomenclature and what to define as strategy or plugin seems a bit odd to me. Methods or approaches seem to be defined inside plugins (ewc, gem, gdumb...) together with the higher hierarchy tiers, such as rehearsal (namely replay). By higher tier I mean that replay should correspond to finetuning with exemplars as a baseline, but can be extended to more methods to create iCaRL, EEIL, BiC and such. Since the number of methods added will grow, I think a better structure would be more beneficial instead of having a mix of the higher hierarchy and the lower (with potential of 50+ methods in a few months). To me, I would consider replay (which is now defined as a subclass of a strategy plugin) and other similar higher hierarchy strategy classes to be in their own folder or be moved to strategies, leaving the methods (the final leaves in the tree of classes) into plugins. On a more personal note, plugins is a name I don't like personally, since I expect plugins to be strategies that can be plugged into multiple methods (like using distillation, storing output statistics...), interchangeable for multiple strategy types even, and not a specific method (like EWC, LwF, ...). Also, a baseline method like joint training is defined in strategies instead of with other methods in plugins. And I see for example how CoPE (from Matthias) is defined with a strategy wrapper, and also in it separate plugin.

Now that some other functionalities that were interfering have been fixed, I was checking in how to continue to port some of the approaches that I have already implemented in the FACIL framework (BiC is the one I talked with Antonio as a start point). The way I do it now is to inherit the replay strategy (inside plugins) and extend it to BiC (also inside plugins). However, I find the way plugins and strategies interlace to be a bit confusing.

**Comment by AntonioCarta** [comment]
Thanks for the feedback @mmasana. Overall, I agree with you that there is a lot of work to do on the Strategy API but I'm not sure how to fix everything yet.

The idea behind plugins was to give an easy mechanism to extend strategies by defining isolated components that can be added to any strategies. These components can be CL strategies themselves, like regularization or replay strategies, that can be more or less considered as separate components that you can add on top of a Naive strategy. But you can also have other kinds of plugins, like a plugin for pytorch learning rate schedulers, or loggers. I believe the main benefit of plugins is the fact that you can split any additional functionality inside its own class without having a large `BaseStrategy` with tons of arguments, with `BaseStrategy` providing only a basic skeleton.

Maybe one of the most confusing points is the fact that currently plugins have access to the entire strategy's state. This makes it possible to always implement a strategy using a plugin instead of subclassing. As a result, some users may doubt whether their strategy should be implemented as a strategy or as a plugin. Maybe restricting plugins could simplify this decision process?

> On a more personal note, plugins is a name I don't like personally, since I expect plugins to be strategies that can be plugged into multiple methods 

Do you have a better name? Remember to consider that plugins are not just strategies, metrics and loggers are also plugins.

> I think a better structure would be more beneficial instead of having a mix of the higher hierarchy and the lower 

This is certainly something that will evolve over time. I'm not sure if a strict hierarchical organization is needed. For example [torch.nn](https://pytorch.org/docs/stable/nn.html) doesn't really have one despite the large number of layers.

Right now we don't have any "higher tier" strategies since they are all defined starting from Naive. However, the `ReplayPlugin` is parametric, which means that you can implement many replay strategies just by changing the `StoragePolicy`. Maybe this is more similar to the kind of organization you would like to have?

> since I expect plugins to be strategies that can be plugged into multiple methods (like using distillation, storing output statistics...), interchangeable for multiple strategy types even, and not a specific method (like EWC, LwF, ...). 

This is something that we should definitely encode and clarify. However, I think some strategies may be considered "pluggable", in the sense that you can attach them to any other strategy. For example, a pure regularization strategy can probably be safely added, like you would do with a regularization function.

**Comment by mmasana** [comment]
> The idea behind plugins was to give an easy mechanism to extend strategies by defining isolated components that can be added to any strategies. These components can be CL strategies themselves, like regularization or replay strategies, that can be more or less considered as separate components that you can add on top of a Naive strategy. But you can also have other kinds of plugins, like a plugin for pytorch learning rate schedulers, or loggers. I believe the main benefit of plugins is the fact that you can split any additional functionality inside its own class without having a large `BaseStrategy` with tons of arguments, with `BaseStrategy` providing only a basic skeleton.

That makes sense, I think the main issue is that if some methods are part plugin and part strategy (or more than one of each), it can become more difficult to understand its implementation. However, I agree that once you get past this initial phase, it can make it easier to combine or extend those parts, or repurpose them to other scenarios. The lack of understanding of how some plugins or strategies work can be detrimental since someone can be bypassing the rules/assumptions of some scenarios. But I guess with proper documentation this can also be mitigated.

> Maybe restricting plugins could simplify this decision process?

My first answer would be yes, since new users could always implement new methods as strategies, and then if something becomes useful, it could be upgraded to plugin with further work on making sure it can be plugged properly in multiple situations (by the avalanche dev team). However, I do understand that would lead to more work and updates on the framework over time. In any case, having one clear way to implement them would always make things easier.

> Do you have a better name? Remember to consider that plugins are not just strategies, metrics and loggers are also plugins.

It's tricky because I would not use those names we are using right now, but I understand the nuances of coming up with smth else. Using words such as method or approach for strategies and the word strategy for plugin would be my go to. But I'm not sure it would solve the problem in some of the cases. Having metrics and loggers into plugins makes sense with the name plugins, but having them mixed with the methods and strategies is quite odd to me, and I wonder how it will look like when 40 new methods are added to the framework.

> This is certainly something that will evolve over time. I'm not sure if a strict hierarchical organization is needed. For example torch.nn doesn't really have one despite the large number of layers.

In this case, I think they have quite a good hierarchy organization with Module() as the basic building block for the network graphs, which then bifurcate into mainly layers and functions that all have the same behaviour and usage. In plugins, you can have metrics and a methods both under the same direct umbrella, which have very different usage.

> This is something that we should definitely encode and clarify. However, I think some strategies may be considered "pluggable", in the sense that you can attach them to any other strategy. For example, a pure regularization strategy can probably be safely added, like you would do with a regularization function.

I agree, but in this example, some plugins are not compatible between them, and I think this is what makes them different. If plugins are different metrics or loggers, you can use as many as you want and they will all provide their output in the corresponding shape. But multiple methods being used as plugins maybe is not so straightforward, and it would be better to fuse them under the strategy umbrella.

**Comment by AntonioCarta** [comment]
I'm adding this discussion to the beta roadmap. I will ping you with a complete answer once we are ready to do the API update.
